### PR TITLE
JAMES-2910 HTML could be indexed directly in ElasticSearch

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/json/MessageToElasticSearchJsonTest.java [hunk 1]
```diff
@@ -385,4 +386,35 @@ public void convertToJsonWithoutAttachmentShouldConvertEmailBoby() throws IOExce
...
+        message.setUid(UID);
+
+        // When
+        MessageToElasticSearchJson messageToElasticSearchJson = new MessageToElasticSearchJson(
+                new JsoupTextExtractor(),
+                ZoneId.of("Europe/Paris"),
+                IndexAttachments.NO);
+        String convertToJsonWithoutAttachment = messageToElasticSearchJson.convertToJsonWithoutAttachment(message, ImmutableList.of(USER));
+
+        System.out.println(convertToJsonWithoutAttachment);
```
Comment (remk) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/json/MessageToElasticSearchJsonTest.java]: debug

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 2]
```diff
@@ -62,17 +62,17 @@ public void setUp() throws Exception {
     }
 
     @Test
-    public void textualContentShouldReturnNullWhenInputStreamIsEmpty() throws Exception {
+    public void textualContentShouldReturnEmptyWhenInputStreamIsEmpty() throws Exception {
         assertThat(textExtractor.extractContent(IOUtils.toInputStream("", StandardCharsets.UTF_8), "text/plain").getTextualContent())
-            .isEmpty();
+            .contains("");
```
Comment (remk) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: equalsTo ?

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 3]
```diff
@@ -77,6 +80,9 @@ private ObjectMapper initializeObjectMapper() {
 
     @Override
     public ParsedContent extractContent(InputStream inputStream, String contentType) throws Exception {
+        if (contentType.startsWith("text/")) {
+            return jsoupTextExtractor.extractContent(inputStream, contentType);
```
Comment (rouazana) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: not only for html?

Comment (chibenwa) [reply mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: Why wiuld we pay the tika price for text?

Comment (rouazana) [reply mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: is jsoup able to handle text only?

Comment (rouazana) [reply mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: you can simply ignore the extract if it is only text, and avoid paying the cost for jsoup also :)

Comment (chibenwa) [reply mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: Well, `you can simply ignore the extract if it is only text,` will result in not getting any text for `text/plain`... Not what we want.

`avoid paying the cost for jsoup also` JsoupTextExtractor parses html only for `text/html` and do avoid html parsing on non HTML document, thus is simply returning the underlying text for non html text.

**Comment by chibenwa** [comment]
```
[36f2f7d1a8ca2f13e1324b998779ed1df8df0dd8] [INFO] # There is insufficient memory for the Java Runtime Environment to continue.
[36f2f7d1a8ca2f13e1324b998779ed1df8df0dd8] [ERROR] # There is insufficient memory for the Java Runtime Environment to continue.OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x0000000633c00000, 484442112, 0) failed; error='Not enough space' (errno=12)
[36f2f7d1a8ca2f13e1324b998779ed1df8df0dd8] [INFO] # Native memory allocation (mmap) failed to map 484442112 bytes for committing reserved memory.
[36f2f7d1a8ca2f13e1324b998779ed1df8df0dd8] [ERROR] # An error report file with more information is saved as:
[36f2f7d1a8ca2f13e1324b998779ed1df8df0dd8] [INFO] # /james-project/hs_err_pid1041.log
```

mvn test :green_apple: 

test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java [hunk 4]
```diff
@@ -43,14 +43,14 @@ public ParsedContent extractContent(InputStream inputStream, String contentType)
...
-           if (contentType.equals("text/plain")) {
-            return new ParsedContent(Optional.ofNullable(IOUtils.toString(inputStream, StandardCharsets.UTF_8)), emptyMetadata);
-           }
-           if (contentType.equals("text/html")) {
-               Document doc = Jsoup.parse(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
-               doc.select(TITLE_HTML_TAG).remove();
-               return new ParsedContent(Optional.ofNullable(doc.text()), emptyMetadata);
-           }
+            if (contentType.equals("text/html")) {
+                Document doc = Jsoup.parse(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java]: There is a parse method using an inputStream which avoid loading the entire html document in memory.

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java [hunk 5]
```diff
@@ -96,7 +97,7 @@ private TextExtractor provideTextExtractor(TikaTextExtractor textExtractor, Tika
         LOGGER.info("Tika text extraction has been disabled." +
             " Using DefaultTextExtractor instead. " +
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java]: please update the log too

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java]: Looks good, thanks!

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java [hunk 6]
```diff
@@ -38,6 +38,9 @@
 
     @Override
     public ParsedContent extractContent(InputStream inputStream, String contentType) throws Exception {
+        if (inputStream == null) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java]: you mix early return and fallback style return.
Could you please use a single style?

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 7]
```diff
@@ -59,12 +60,14 @@
...
     private final TikaHttpClient tikaHttpClient;
     private final ObjectMapper objectMapper;
+    private final JsoupTextExtractor jsoupTextExtractor;
 
     @Inject
     public TikaTextExtractor(MetricFactory metricFactory, TikaHttpClient tikaHttpClient) {
         this.metricFactory = metricFactory;
         this.tikaHttpClient = tikaHttpClient;
         this.objectMapper = initializeObjectMapper();
+        jsoupTextExtractor = new JsoupTextExtractor();
```
Comment (mbaechler) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: you lack `this` for consistency

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 8]
```diff
@@ -62,17 +62,17 @@ public void setUp() throws Exception {
...
     }
 
     @Test
     public void textTest() throws Exception {
         InputStream inputStream = ClassLoader.getSystemResourceAsStream("documents/Text.txt");
         assertThat(inputStream).isNotNull();
         assertThat(textExtractor.extractContent(inputStream, "text/plain").getTextualContent())
-            .contains("This is some awesome text text.\n\n\n");
+            .contains("This is some awesome text text.\n\n");
```
Comment (mbaechler) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: do we care about the number of `\n`?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java [hunk 9]
```diff
@@ -43,14 +43,14 @@ public ParsedContent extractContent(InputStream inputStream, String contentType)
...
-               return new ParsedContent(Optional.ofNullable(doc.text()), emptyMetadata);
-           }
+            if (contentType.equals("text/html")) {
+                Document doc = Jsoup.parse(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
+                doc.select(TITLE_HTML_TAG).remove();
+                return new ParsedContent(Optional.ofNullable(doc.text()), emptyMetadata);
+            }
+            if (contentType.startsWith("text/")) {
+                return new ParsedContent(Optional.ofNullable(IOUtils.toString(inputStream, StandardCharsets.UTF_8)), emptyMetadata);
+            }
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java]: are you sure you want to parse `text/calendar` with `JsoupTextExtractor`?

**Comment by chibenwa** [comment]
MemorySpamAssassinContractTest.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations » ConditionTimeout

test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java [hunk 10]
```diff
@@ -31,28 +32,37 @@
...
-            if (contentType.equals("text/html")) {
-                Document doc = Jsoup.parse(inputStream, StandardCharsets.UTF_8.name(), NO_BASE_URI);
-                doc.select(TITLE_HTML_TAG).remove();
-                return new ParsedContent(Optional.ofNullable(doc.text()), emptyMetadata);
-            }
-            if (contentType.equals("text/plain")) {
-                return new ParsedContent(Optional.ofNullable(IOUtils.toString(inputStream, StandardCharsets.UTF_8)), emptyMetadata);
-            }
+        if (contentType == null) {
+            return ParsedContent.empty();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extractor/JsoupTextExtractor.java]: it still returns this case twice

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 11]
```diff
@@ -72,85 +72,90 @@ public void textTest() throws Exception {
...
     }
     
     @Test
     public void excelTest() throws Exception {
         InputStream inputStream = ClassLoader.getSystemResourceAsStream("documents/calc.xlsx");
         assertThat(inputStream).isNotNull();
         assertThat(textExtractor.extractContent(inputStream, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet").getTextualContent())
+            .isPresent()
+            .asString()
             .contains("Feuille1\n" +
```
Comment (mbaechler) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: you can have several contains, don't you?

**Comment by chibenwa** [comment]
```
[1c0f97530f51eb950a44300e32f2a748e7a940e5] [ERROR] sendingOneHundredMessagesShouldBeRandomlyAssignedToEveryMailboxesOfEveryUsers  Time elapsed: 323.578 s  <<< ERROR!
[1c0f97530f51eb950a44300e32f2a748e7a940e5] org.awaitility.core.ConditionTimeoutException: 
[1c0f97530f51eb950a44300e32f2a748e7a940e5] Assertion condition defined as a lambda expression in org.apache.james.smtp.SmtpRandomStoringTest 
[1c0f97530f51eb950a44300e32f2a748e7a940e5] Expecting:
[1c0f97530f51eb950a44300e32f2a748e7a940e5]  <39L>
[1c0f97530f51eb950a44300e32f2a748e7a940e5] to be between:
[1c0f97530f51eb950a44300e32f2a748e7a940e5]  [400L, 800L] within 10 seconds.
[1c0f97530f51eb950a44300e32f2a748e7a940e5] 	at org.apache.james.smtp.SmtpRandomStoringTest.sendingOneHundredMessagesShouldBeRandomlyAssignedToEveryMailboxesOfEveryUsers(SmtpRandomStoringTest.java:156)
[1c0f97530f51eb950a44300e32f2a748e7a940e5] Caused by: java.util.concurrent.TimeoutException
[1c0f97530f51eb950a44300e32f2a748e7a940e5] 	at org.apache.james.smtp.SmtpRandomStoringTest.sendingOneHundredMessagesShouldBeRandomlyAssignedToEveryMailboxesOfEveryUsers(SmtpRandomStoringTest.java:156)
[1c0f97530f51eb950a44300e32f2a748e7a940e5] 
```

test this please

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 12]
```diff
@@ -154,8 +156,8 @@ public void excelTest() throws Exception {
         assertThat(textExtractor.extractContent(inputStream, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet").getTextualContent())
             .isPresent()
             .asString()
-            .contains("Feuille1\n" +
-                "\tThis is an aesome LibreOffice document !");
+            .contains("Feuille1")
+            .contains("This is an aesome LibreOffice document !");
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: Won't fix as the typo is also in the referenced LibreOffice document

**Comment by chibenwa** [comment]
The email corpus we tested this against does not contains HTML nor attachment and thus is not significative for performance evaluation of this change set.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**